controllers/pod_node_module_reconciler.go

```diff
@@ -0,0 +1,166 @@
+	}
+
+	nodeName := pod.Spec.NodeName
+
+	moduleName, ok := pod.Labels[constants.ModuleNameLabel]
+	if !ok {
+		return ctrl.Result{}, fmt.Errorf("pod %s has no %q label", podNamespacedName, constants.ModuleNameLabel)
+	}
+
+	labelName := daemonset.GetDriverContainerNodeLabel(moduleName)
+
+	deleting := !pod.DeletionTimestamp.IsZero()
+
+	logger = logger.WithValues(
+		"node name", nodeName,
+		"module name", moduleName,
+		"label name", labelName,
+		"deleting pod", deleting,
+	)
+
+	if !podutils.IsPodReady(&pod) || deleting {
+		logger.Info("Unlabeling node")
+
+		if err := pnmr.deleteLabel(ctx, nodeName, labelName); err != nil {
+			return ctrl.Result{}, fmt.Errorf("could not unlabel node %s: %v", nodeName, err)
+		}
+
+		if deleting {
+			logger.Info("Removing finalizer")
+
+			if err := pnmr.deleteFinalizer(ctx, &pod); err != nil {
+				return ctrl.Result{}, fmt.Errorf("could not delete the pod finalizer: %v", err)
+			}
+		}
+
+		return ctrl.Result{}, nil
+	} else {
+		logger.Info("Labeling node")
+
+		return ctrl.Result{}, pnmr.addLabel(ctx, nodeName, labelName)
+	}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (pnmr *PodNodeModuleReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	p := predicate.And(
+		PodReadinessChangedPredicate(
+			mgr.GetLogger().WithName("pod-readiness-changed"),
+		),
+		filter.HasLabel(constants.ModuleNameLabel),
+	)
+
+	return ctrl.
+		NewControllerManagedBy(mgr).
+		Named("pod-node-module").
+		For(&v1.Pod{}).
+		WithEventFilter(p).
+		Complete(pnmr)
+}
+
+func (pnmr *PodNodeModuleReconciler) addLabel(ctx context.Context, nodeName, labelName string) error {
+	node := v1.Node{}
+
+	if err := pnmr.client.Get(ctx, types.NamespacedName{Name: nodeName}, &node); err != nil {
+		return fmt.Errorf("could not get node %s: %v", nodeName, err)
+	}
+
+	nodeCopy := node.DeepCopy()
+
+	if node.Labels == nil {
+		node.Labels = make(map[string]string, 1)
+	}
+
+	node.Labels[labelName] = ""
+
+	return pnmr.client.Patch(ctx, &node, client.MergeFrom(nodeCopy))
+}
+
+func (pnmr *PodNodeModuleReconciler) deleteFinalizer(ctx context.Context, pod *v1.Pod) error {
+	podCopy := pod.DeepCopy()
+
+	controllerutil.RemoveFinalizer(pod, constants.NodeLabelerFinalizer)
+
+	return pnmr.client.Patch(ctx, pod, client.MergeFrom(podCopy))
+}
+
+func (pnmr *PodNodeModuleReconciler) deleteLabel(ctx context.Context, nodeName, labelName string) error {
+	node := v1.Node{}
+
+	if err := pnmr.client.Get(ctx, types.NamespacedName{Name: nodeName}, &node); err != nil {
+		return fmt.Errorf("could not get node %s: %v", nodeName, err)
+	}
+
+	nodeCopy := node.DeepCopy()
+
+	delete(node.Labels, labelName)
+
+	return pnmr.client.Patch(ctx, &node, client.MergeFrom(nodeCopy))
+}
+
+// PodReadinessChangedPredicate returns a predicate for Update events that only returns true if the Ready condition
+// changed.
+func PodReadinessChangedPredicate(logger logr.Logger) predicate.Predicate {
+	return predicate.Funcs{
```

Comment:
what do you think about moving all the predicates into the filter package? eventually this is what this package is for

Reply:
I am a bit torn on that one. It makes sense to have them all in the same place, but that predicate is quite specific and I think having it close to the reconciler makes the logic easier to understand.

Reply:
I would prefer to have them all concentrated in one place. will allow optimizations of the code if need be, and makes it more clear to understand